zkevm-circuits/src/bin/stats/halo2_stats.rs

```diff
@@ -0,0 +1,374 @@
+    // drop(advice_lagranges);
+    // ```
+    //
+    //
+    // ### 8. Generate necessary extended lagranges.
+    //
+    // M.C = 2
+    // M.S = 1 + 2 * c_f + 2 * c_p + 2 * c_i + c_a + 3 * c_l + c_pg
+    // M.E = 3 + c_f + c_p + c_pg + c_i + c_a (+= c_i + c_a)
+    // ```
+    // let instances_extended_lagrnages = instances_monomials.to_extended_lagranges();
+    // let advice_extended_lagrnages = advice_monomials.to_extended_lagranges();
+    // ```
+    //
+    //
+    // ### 9. While computing the quotient, these extended lagranges:
+    //
+    // ⬡ permuted_input_extended_lagranges
+    // ⬡ permuted_table_extended_lagranges
+    // ⬡ lookup_product_extended_lagranges
+    //
+    // of each lookup argument are generated on the fly and drop before next.
+    //
+    // And 1 extra quotient_extended_lagrange is created. So the peak memory:
+    //
+    // M.C = 2
+    // M.S = 1 + 2 * c_f + 2 * c_p + 2 * c_i + c_a + 3 * c_l + c_pg
+    // M.E = 4 + c_f + c_p + c_pg + c_i + c_a + 3 * (c_l > 0) (+= 3 * (c_l > 0) + 1)
+    // ```
+    // let quotient_extended_lagrange = quotient_extended_lagrange();
+    // ```
+    //
+    //
+    // ### 10. After quotient is comuputed, drop all the other extended lagranges.
+    //
+    // M.C = 2
+    // M.S = 1 + 2 * c_f + 2 * c_p + 2 * c_i + c_a + 3 * c_l + c_pg
+    // M.E = 4 + c_f + c_p (-= c_pg + c_i + c_a + 3 * (c_l > 0))
+    // drop(instances_extended_lagrnages)
+    // drop(advice_extended_lagrnages)
+    // drop(perm_grand_product_extended_lagranges)
+    //
+    //
+    // ### 11. Turn quotient_extended_lagrange into monomial form.
+    // And then cut int `d - 1` pieces.
+    //
+    // M.C = 2
+    // M.S = 2 * c_f + 2 * c_p + 2 * c_i + c_a + 3 * c_l + c_pg + d (+= d - 1)
+    // M.E = 3 + c_f + c_p (-= 1)
+    // ```
+    // let quotient_monomials = quotient_monomials()
+    // drop(quotient_extended_lagrange)
+    // ```
+    //
+    //
+    // ### 12. Evaluate and open all polynomial except instance ones.
+    // }
+    pub(crate) fn estimate_peak_mem(&self, k: u32) -> usize {
+        let field_bytes = 32;
+        let c_f = self.num_fixed_columns;
+        let c_a = self.num_advice_columns;
+        let c_i = self.num_instance_columns;
+        let c_p = self.num_permutation_columns;
+        let c_l = self.num_lookups;
+        let c_pg = (c_p + self.degree - 3) / (self.degree - 2);
+        let e = (self.degree - 1).next_power_of_two();
+        // The estimated peak memory formula comes from point 9 of the analysis, which is the step
+        // of proving that needs the most memory (after that step, allocations start getting freed)
+
+        // number of "elliptic curve points"
+        let m_c = 2;
+        // number of "field elements"
+        let m_s = 1 + 2 * c_f + 2 * c_p + 2 * c_i + c_a + 3 * c_l + c_pg;
+        // number of "field elements" that will be extended by "* e"
+        let m_e = 4 + c_f + c_p + c_pg + c_i + c_a + 3 * (c_l > 0) as usize;
```

Comment:
These formulas for `m_s` and `m_e` are from step 10. Should they be from step 11?

Reply:
The formulas for `m_s` and `m_e` should correspond to step 9 (which is the step that has the peack memory usage).  I think the current code is correct:
https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/7020216f02985a29232df398032b94d4b8d6109b/zkevm-circuits/src/bin/stats/halo2_stats.rs#L190-L202